File: 918.环形子数组的最大和/918.环形子数组的最大和.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& A) {
        int len = A.size();
        vector<int> dp(len);
        dp[0]=A[0];
        int sum = A[0];//整个数组的和
        int maxSum = INT_MIN;//最大子序和
        for(int i=1;i<len;i++){
            sum += A[i];
            dp[i] = max(dp[i-1]+A[i],A[i]);
            maxSum = max(maxSum,dp[i]);
        }
        int minSum = 0;// A[1]-A[n-1]的最小子序和
        for(int i =1;i<len-1;i++){
            dp[i] = min(dp[i-1]+A[i],A[i]);
            minSum = min(minSum,dp[i]);
        }
        return max(sum - minSum, maxSum);
    }
};

File: Solution/918.环形子数组的最大和/918.环形子数组的最大和.java
/*
 * @Descroption: LeetCode 918. 环形子数组的最大和
 * @Author: EmoryHuang
 * @Date: 2023-07-20 09:39:56
 * @Method:
 * 动态规划
 * 分别考虑取到头尾和不取到头尾的情况
 */

import java.util.*;

class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int n = nums.length;
        int preMax = nums[0], sumMax = nums[0];
        int preMin = nums[0], sumMin = nums[0];
        int sum = nums[0];
        for (int i = 1; i < n; i++) {
            preMax = Math.max(preMax + nums[i], nums[i]);
            sumMax = Math.max(sumMax, preMax);
            preMin = Math.min(preMin + nums[i], nums[i]);
            sumMin = Math.min(sumMin, preMin);
            sum += nums[i];
        }
        if (sumMax < 0) {
            return sumMax;
        } else {
            return Math.max(sumMax, sum - sumMin);
        }
    }
}